examples/toplevel-decoration.c

```diff
@@ -0,0 +1,189 @@
+		uint32_t name, const char *interface, uint32_t version) {
+	if (strcmp(interface, "wl_compositor") == 0) {
+		compositor = wl_registry_bind(registry, name, &wl_compositor_interface,
+			1);
+	} else if (strcmp(interface, "xdg_wm_base") == 0) {
+		wm_base = wl_registry_bind(registry, name, &xdg_wm_base_interface, 1);
+	} else if (strcmp(interface, "zxdg_toplevel_decoration_manager_v1") == 0) {
+		decoration_manager = wl_registry_bind(registry, name,
+			&zxdg_toplevel_decoration_manager_v1_interface, 1);
+	}
+}
+
+static void handle_global_remove(void *data, struct wl_registry *registry,
+		uint32_t name) {
+	// TODO
+}
+
+static const struct wl_registry_listener registry_listener = {
+	.global = handle_global,
+	.global_remove = handle_global_remove,
+};
+
+int main(int argc, char **argv) {
+	if (argc == 2) {
+		char *mode = argv[1];
+		if (strcmp(mode, "client") == 0) {
+			decoration_mode = ZXDG_TOPLEVEL_DECORATION_V1_MODE_CLIENT;
+		} else if (strcmp(mode, "server") == 0) {
+			decoration_mode = ZXDG_TOPLEVEL_DECORATION_V1_MODE_SERVER;
+		} else {
+			fprintf(stderr, "Invalid decoration mode\n");
+			return EXIT_FAILURE;
+		}
+	}
+
+	struct wl_display *display = wl_display_connect(NULL);
+	if (display == NULL) {
+		fprintf(stderr, "Failed to create display\n");
+		return EXIT_FAILURE;
+	}
+
+	struct wl_registry *registry = wl_display_get_registry(display);
+	wl_registry_add_listener(registry, &registry_listener, NULL);
+	wl_display_dispatch(display);
+	wl_display_roundtrip(display);
+
+	if (compositor == NULL) {
+		fprintf(stderr, "wl-compositor not available\n");
+		return EXIT_FAILURE;
+	}
+	if (wm_base == NULL) {
+		fprintf(stderr, "xdg-shell not available\n");
+		return EXIT_FAILURE;
+	}
+	if (decoration_manager == NULL) {
+		fprintf(stderr, "xdg-toplevel-decoration not available\n");
+		return EXIT_FAILURE;
+	}
+
+	wlr_egl_init(&egl, EGL_PLATFORM_WAYLAND_EXT, display, NULL,
+		WL_SHM_FORMAT_ARGB8888);
+
+	struct wl_surface *surface = wl_compositor_create_surface(compositor);
+	struct xdg_surface *xdg_surface =
+		xdg_wm_base_get_xdg_surface(wm_base, surface);
+	struct xdg_toplevel *xdg_toplevel = xdg_surface_get_toplevel(xdg_surface);
+	struct zxdg_toplevel_decoration_v1 *decoration =
+		zxdg_toplevel_decoration_manager_v1_get_decoration(decoration_manager,
+		xdg_toplevel);
+
+	if (decoration_mode != 0) {
+		zxdg_toplevel_decoration_v1_set_mode(decoration, decoration_mode);
+	}
+
+	xdg_surface_add_listener(xdg_surface, &xdg_surface_listener, NULL);
+	xdg_toplevel_add_listener(xdg_toplevel, &xdg_toplevel_listener, NULL);
+	zxdg_toplevel_decoration_v1_add_listener(decoration, &decoration_listener,
+		NULL);
+	wl_surface_commit(surface);
+
+	egl_window = wl_egl_window_create(surface, width, height);
+	egl_surface = wlr_egl_create_surface(&egl, egl_window);
+
+	wl_display_roundtrip(display);
+
+	draw();
+
+	while (wl_display_dispatch(display) != -1) {
+		// No-op
+	}
+
+	return EXIT_SUCCESS;
+}
```

Comment:
Would be nice if the example could be made to toggle CSD at runtime

Reply:
Just click on the window, it will switch the mode.